### ISSUE TITLE
Compilation error for `impl` for another const generic trait

### ISSUE THREAD

**Issue by tats-u** [rust-lang/rust#100405]
Given the following code: 

```rust
pub trait ExtraFieldHeader {
    fn header_id(&self) -> u16;
}

pub trait FixedIDZIPExtraField<const ID: u16> {
    const HEADER_ID: u16 = ID;
    /// Gets header ID (Unique for each type of extra fields)
    fn header_id(&self) -> u16 {
        Self::HEADER_ID
    }
}

impl<T, const ID: u16> ExtraFieldHeader for T where T: FixedIDZIPExtraField<ID> {}
```

The current output is:

```
error[E0207]: the const parameter `ID` is not constrained by the impl trait, self type, or predicates
  --> src\zip_extra_field.rs:27:9
   |
27 | impl<T, const ID: u16> ExtraFieldHeader for T where T: FixedIDZIPExtraField<ID> {}
   |         ^^^^^^^^^^^^^ unconstrained const parameter
   |
   = note: expressions using a const parameter must map each value to a distinct output value
   = note: proving the result of expressions other than the parameter are unique is not supported
```

Ideally the output should look like:

```
(no errors)
```

I believe `ID` is constrained for `FiexedIDZIPExtraField<ID>` and this error must not be raised. Could someone provide a solution?

**Comment by tats-u** [comment]
Seems to be similar to https://github.com/rust-lang/rust/issues/100405

**Comment by compiler-errors** [comment]
@tats-u did you mean to link another issue?

**Comment by tats-u** [comment]
@compiler-errors I just thought this is similar to that issue.
Do you think this is duplicated?

**Comment by compiler-errors** [comment]
You said that this issue (#100405) is similar to issue #100405... which is this one?

**Comment by tats-u** [comment]
@compiler-errors Sorry, I thought I copied and pasted another issue...

**Comment by tats-u** [comment]
What I have wanted to mention are https://github.com/rust-lang/rust/issues/68366 or https://github.com/rust-lang/rust/issues/81985.

**Comment by aDogCalledSpot** [comment]
This doesn't appear to be unique to const generics: 

```rust
trait Foo<T> {}

impl<T, U: Foo<T>> Drop for U {
    fn drop(&mut self) {
        println!("hello");
    }
}
```

leads to:

```
error[E0207]: the type parameter `T` is not constrained by the impl trait, self type, or predicates
 --> src/main.rs:3:6
  |
3 | impl<T, U: Foo<T>> Drop for U {
  |      ^ unconstrained type parameter

For more information about this error, try `rustc --explain E0207`.
```

Using `rustc 1.67.0-nightly (53e4b9dd7 2022-12-04)`.

**Comment by ZER0** [comment]
Is there any workaround on this until is got (hopefully) fixed…?

**Comment by BaxHugh** [comment]
I've also got this problem, has any progress been made with this?
